refine cmake files:

dfukalov opened the pull request:
1. use add_subdirectory to include llvm/clang tree instead of add_custom_target.
   there were explicit cmake&make invokes for each hcc build
2. version string building code moved out to function add_version_info_from_git
3. removed confusing instruction to do "make world" for first time
4. removed a number of not used variables (like EXECUTABLE_OUTPUT_PATH)
5. removed use of ProcessorCount() since the llvm tree make is a part of whole
   hcc build
6. removed using CMAKE_FORCE_CXX_COMPILER - it is deprecated in cmake 3.6

dfukalov committed:
refine cmake files:
1. use add_subdirectory to include llvm/clang tree instead of add_custom_target.
   there were explicit cmake&make invokes for each hcc build
2. version string building code moved out to function add_version_info_from_git
3. removed confusing instruction to do "make world" for first time
4. removed a number of not used variables (like EXECUTABLE_OUTPUT_PATH)
5. removed use of ProcessorCount() since the llvm tree make is a part of whole
   hcc build
6. removed using CMAKE_FORCE_CXX_COMPILER - it is deprecated in cmake 3.6

pfultz2 commented on cmake-tests/CMakeLists.txt:
```diff
@@ -1,8 +1,9 @@
-endif()
 add_executable(cmake-test cmake-test.cpp)
+
+add_dependencies(cmake-test world)
```
This breaks the test when ran as standalone, which is useful for testing the `find_package(hcc)` logic. Either this line should be moved into the `if(TARGET hccrt)` conditional block or hccrt should be update to depend on the world target, so this line can be removed.

scchan replied:
@dfukalov could you address this?

dfukalov replied:
moving `add_dependencies(cmake-test world)` into `if(TARGET hccrt)` block doesn't work for me since cmake-test binary wants .so and .bc files in right places. I'm reworking the change now

pfultz2 commented on CMakeLists.txt:
```diff
@@ -325,29 +287,19 @@ set(KALMAR_BACKEND "HCC_BACKEND_AMDGPU")
-)
+
+set(LLVM_TARGETS_TO_BUILD "AMDGPU" "X86"
+    CACHE STRING "Semicolon-separated list of targets to build, or \"all\"." FORCE)
```
Is using `FORCE` necessary here? It seems this won't allow the user to choose different targets from the command-line.

dfukalov replied:
you're right, I'll remove it

scchan reviewed (changes requested):
Please address concerns raised by @pfultz2

dfukalov committed:
modified according to Paul' review notes

dfukalov commented:
I've just updated the change according to Paul' comments. `cmake-test` dependency of `world` target moved into `hccrt` case it works for `all` build.
`make cmake-test` in clean hcc' build directory fails, I'm going to fix it in separate change since it'll modify more files

pfultz2 commented:
> make cmake-test in clean hcc' build directory fails,

Why does that fail?

dfukalov commented:
this target at least uses clamp-device script that invokes opt and llc and they uses our passes from .so files and all of them are located in install bin directory.
but cmake-test target in hcc build tree has no dependency from these (llc opt etc.) targets

dfukalov commented:
Paul, I've just prepared next change that fixes cmake-test target clean build. Is this branch ready to land? Or I should merge cmake-test fix into the change?

pfultz2 commented:
> Paul, I've just prepared next change that fixes cmake-test target clean build. Is this branch ready to land? Or I should merge cmake-test fix into the change?

If you have the cmake-test fix ready, then why not?

whchung reviewed (approved):
LGTM. @scchan ?

dfukalov committed:
fixed cmake-test standalone build:
1. clamp scripts generation from .in files moved from tests/CMakeLists.txt
   to lib/CMakeLists.txt
2. added custom target clamp-scripts
3. cmake-test target dependency reduced from world to clamp-scripts,
   clang_links, builtins
